sites/wcpl.test/crontab_settings.txt
# Edit this file to introduce tasks to be run by cron.
#
# Each task to run has to be defined through a single line
# indicating with different fields when the task will be run
# and what command to run for the task
#
# To define the time you can provide concrete values for
# minute (m), hour (h), day of month (dom), month (mon),
# and day of week (dow) or use '*' in these fields (for 'any').#
# Notice that tasks will be started based on the cron's system
# daemon's notion of time and timezones.
#
# Output of the crontab jobs (including errors) is sent through
# email to the user the crontab file belongs to (unless redirected).
#
# For example, you can run a backup of all your user accounts
# at 5 a.m every week with:
# 0 5 * * 1 tar -zcf /var/backups/home.tgz /home/
#
# For more information see the manual pages of crontab(5) and cron(8)
#
# m h  dom mon dow   command
# cleanup of tmp files and restart solr to clean up memory

###################
## Crontab setting from a 'configuration' file
## 
## to save current crontab to file: crontab -l > crontab_settings.txt
##
## to set the current crontab from a file : crontab < crontab_settings.txt
##
## added by pascal. 12-11-2014
##################

# !! Please make all changes to the crontab within the git repository system and apply through the above method.
#
# ie :
# git pull origin {current branch}
# sudo crontab < /usr/local/vufind-plus/sites/wcpl.test/conf/crontab_settings.txt

###############
# On Reboot
###############
@reboot cd /usr/local/vufind-plus/sites/wcpl.test; ./wcpl.test start
# restart continuous re-indexing on reboot.
@reboot cd /usr/local/vufind-plus/sites/wcpl.test; ./continuous_partial_reindex.sh

#######################
# VuFind-Plus Cleanup #
#######################
27 0 * * * cd /tmp; rm -rf CURLCOOKIE*
28 0 * * * cd /tmp; rm -rf epb*

########################
# Full Update #
########################
15 5 * * * cd /usr/local/vufind-plus/sites/wcpl.test; ./full_update.sh

######################
# MySQL Nightly Dump #
######################
# backup important bits at 12:10am daily
10 0 * * * /usr/local/vufind-plus/vufind/cron/nightly_mysql_dump.sh wcpl.test 2>&1 >/dev/null

##########
# Cron   #
##########
5 */3 * * * cd /usr/local/vufind-plus/vufind/cron/; java -jar cron.jar wcpl.test
# run pika cron jobs every three hours

20 0 * * * cd /usr/local/vufind-plus/vufind/cron/; java -jar cron.jar wcpl.test DatabaseCleanup

# Check Apache Solr process and website that it has not stalled.
*/3 * * * * cd /usr/local/vufind-plus/vufind/cron/; ./solrTest.sh


sites/default/conf/crontab_settings.txt
###################
## Crontab setting from a 'configuration' file
##
## to save current crontab to file: crontab -l > crontab_settings.txt
##
## to set the current crontab from a file : crontab < crontab_settings.txt
##
##################

# !! Please make all changes to the crontab within the git repository system and apply through the above method.
#
# ie :
# git pull origin {current branch}
# crontab < /usr/local/vufind-plus/sites/{your_server_name}/conf/crontab_settings.txt

###################
# * * * * *  command to execute
# │ │ │ │ │
# │ │ │ │ └───── day of week (0 - 6)
# │ │ │ │        (0 to 6 are Sunday to Saturday, or use names (Sun,Mon,Tue,Wed,Thu,Fri,Sat); 7 is Sunday, the same as 0)
# │ │ │ └────────── month (1 - 12)
# │ │ └─────────────── day of month (1 - 31)
# │ └──────────────────── hour (0 - 23)
# └───────────────────────── min (0 - 59)
## taken from https://en.wikipedia.org/wiki/Cron
##############

# Below are common cron tasks used in Pika setups.
# Uncomment (& adjust timing/settings) to implement the cron tasks your installation needs
#

#############
# On Reboot #
#############
# Continuous Re-Indexing
#@reboot cd /usr/local/vufind-plus/sites/{your_server_name}; ./extract_and_reindex_then_email.sh

################
# Pika Cleanup #
################

# Solr Restart
#7 1 * * * cd /usr/local/vufind-plus/sites/{your_server_name}; ./{your_server_name}.sh restart
# alternate to above, if pika is setup as a system service
#7 1 * * * service pika restart

# Clean Up Temporary Files
#26 * * * * cd /tmp; /bin/rm -rf CURLCOOKIE*
#27 */2 * * * cd /tmp; /bin/rm -rf epb*

###################################
# Data Extract from other systems #
###################################
# OverDrive Data - every 15 minutes throughout the day for near runtime availability
#2-59/15 0,3-23 * * * cd /usr/local/vufind-plus/vufind/overdrive_api_extract/; nice -n -10 java -jar overdrive_extract.jar {your_server_name}

# Hoopla extract - once a week for the entire server
#5 23 * * 0 cd /usr/local/vufind-plus/vufind/cron;./HOOPLA.sh {your_server_name}

########################
# Update Grouped Works #
########################
# Do a full regrouping once a day in the middle the night, but not during the day
#10 0 * * * cd /usr/local/vufind-plus/vufind/record_grouping; java -server -Xmx6G -XX:+UseParallelGC -XX:ParallelGCThreads=2 -jar record_grouping.jar {your_server_name} fullRegroupingNoClear

###########
# Reindex #
###########
# Do a full reindex at night after the works have been updated
#10 1 * * * cd /usr/local/vufind-plus/vufind/reindexer; nice -n -3 java -jar reindexer.jar {your_server_name} fullReindex

# Do a partial reindex every 15 minutes during the day after extract for near runtime availability
# Skip doing updates from 1-3 though since we are doing a full reindex
#7-59/15 0,3-23 * * * cd /usr/local/vufind-plus/vufind/reindexer; nice -n -5 java -jar reindexer.jar {your_server_name}

###################
# Pika Cron Jobs  #
###################
#7 */3 * * * cd /usr/local/vufind-plus/vufind/cron/; java -jar cron.jar {your_server_name}
20 0 * * * cd /usr/local/vufind-plus/vufind/cron/; java -jar cron.jar {your_server_name} DatabaseCleanup

# Check Apache Solr process and website that it has not stalled.
*/3 * * * * cd /usr/local/vufind-plus/vufind/cron/; ./solrTest.sh

vufind/web/services/Admin/Config.php
<?php
/**
 * $Id: Config.php 1967 2009-12-18 15:40:23Z demiankatz $
 *
 * Copyright (C) Villanova University 2007.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

require_once ROOT_DIR . '/Action.php';
require_once ROOT_DIR . '/sys/ConfigArray.php';
require_once ROOT_DIR . '/services/Admin/Admin.php';

class Admin_Config extends Admin_Admin
{
	function launch()
	{
		global $configArray;
		global $interface;

		if (isset($_POST['submit'])) {
			switch ($_GET['file']) {
				case 'stopwords.txt':
					$this->processStopWords();
					break;
				case 'synonyms.txt':
					$this->processSynonyms();
					break;
				case 'protwords.txt':
					$this->processProtWords();
					break;
				case 'elevate.xml':
					//$this->processElevate();
					$this->processConfigFile($configArray['Index']['local'] . '/biblio/conf/elevate.xml',
					$_POST['config_file']);
					break;
				case 'searchspecs.yaml':
				case 'searches.ini':
				case 'facets.ini':
					// Figure out the path to the file based on its extension -- the YAML option is
					// hard-coded, but the INI files may be renamed by the user.
					list($filename, $ext) = explode('.', $_GET['file']);
					if ($ext == 'yaml') {
						$path = 'conf/' . $_GET['file'];
					} else {
						$path = getExtraConfigArrayFile($filename);
					}
					$this->processConfigFile($configArray['Site']['local'] . '/' . $path, $_POST['config_file']);
					break;
				case 'config.ini':
					//$this->processConfig();   // dangerous -- disabled until fixed
					$this->processConfigFile($configArray['Site']['local'] . '/' . 'conf/config.ini',
					$_POST['config_file']);
					break;
			}
		}

		if (isset($_GET['file'])) {
			switch ($_GET['file']) {
				case 'stopwords.txt':
					$this->showStopWords();
					break;
				case 'synonyms.txt':
					$this->showSynonyms();
					break;
				case 'protwords.txt':
					$this->showProtWords();
					break;
				case 'elevate.xml':
					//$this->showElevate();
					$this->showConfigFile('Elevated Terms Configuration',
					$configArray['Index']['local'] . '/biblio/conf/elevate.xml');
					break;
				case 'searchspecs.yaml':
					$this->showConfigFile('Search Specifications',
					$configArray['Site']['local'] . '/' . 'conf/searchspecs.yaml');
					break;
				case 'searches.ini':
					$this->showConfigFile('Search Settings',
					$configArray['Site']['local'] . '/' . getExtraConfigArrayFile('searches'));
					break;
				case 'facets.ini':
					$this->showConfigFile('Facet Settings',
					$configArray['Site']['local'] . '/' . getExtraConfigArrayFile('facets'));
					break;
				case 'config.ini':
				default:
					//$this->showConfig();   // dangerous -- disabled until fixed
					$this->showConfigFile('General Settings',
					$configArray['Site']['local'] . '/' . 'conf/config.ini');
					break;
			}
		} else {
			$interface->setPageTitle('Configuration');
			$interface->setTemplate('config.tpl');
		}
		$interface->display('layout.tpl');
	}

	function showStopWords()
	{
		global $interface;
		global $configArray;

		$stopwords = file_get_contents($configArray['Index']['local'] . '/biblio/conf/stopwords.txt');
		$interface->assign('stopwords', $stopwords);
		$interface->setPageTitle("Stop Words Configuration");
		$interface->setTemplate('config-stopwords.tpl');
	}

	function processStopWords()
	{
		global $configArray;

		$this->processConfigFile($configArray['Index']['local'] . '/biblio/conf/stopwords.txt',
		$_POST['stopwords']);
	}

	function showSynonyms()
	{
		global $interface;
		global $configArray;

		$synonyms = file_get_contents($configArray['Index']['local'] . '/biblio/conf/synonyms.txt');
		$interface->assign('synonyms', $synonyms);
		$interface->setPageTitle("Synonyms Configuration");
		$interface->setTemplate('config-synonyms.tpl');
	}

	function processSynonyms()
	{
		global $configArray;

		$this->processConfigFile($configArray['Index']['local'] . '/biblio/conf/synonyms.txt',
		$_POST['synonyms']);
	}

	function showProtWords()
	{
		global $interface;
		global $configArray;

		$protwords = file_get_contents($configArray['Index']['local'] . '/biblio/conf/protwords.txt');
		$interface->assign('protwords', $protwords);
		$interface->setPageTitle("Protected Words Configuration");
		$interface->setTemplate('config-protwords.tpl');
	}

	function processProtWords()
	{
		global $configArray;

		$this->processConfigFile($configArray['Index']['local'] . '/biblio/conf/protwords.txt',
		$_POST['protwords']);
	}

	/* Not fully implemented -- commented out until this can be finished; in the meantime, we'll
	 * use the generic showConfigFile / processConfigFile methods.
	 function showElevate()
	 {
	 global $interface;
	 global $configArray;

	 $elevate = file_get_contents($configArray['Index']['local'] . '/biblio/conf/elevate.xml');
	 $interface->assign('elevate', $elevate);
	 $interface->setPageTitle("Elevated Terms Configuration");
	 $interface->setTemplate('config-elevate.tpl');
	 }

	 function processElevate()
	 {
	 global $configArray;

	 $doc = new DOM_Document();
	 $xml = $doc->saveXML();

	 $this->processConfigFile($configArray['Index']['local'] . '/biblio/conf/elevate.xml',
	 $xml);
	 }
	 */

	function showConfigFile($title, $path)
	{
		global $interface;

		$file = @file_get_contents($path);
		$interface->assign('configPath', $path);
		$interface->assign('configFile', $file);
		$interface->setPageTitle($title);
		$interface->setTemplate('config-file.tpl');
	}

	function processConfigFile($filename, $contents)
	{
		global $interface;

		$interface->assign('saved', true);
		$interface->assign('bytesWritten', @file_put_contents($filename, $contents));
	}

	/* This code is incomplete and dangerous since it can corrupt config.ini --
	 * it's disabled for now until we can finish it.

	 function showConfig()
	 {
	 global $interface;
	 global $configArray;

	 $list = array();
	 $themesDir = $configArray['Site']['local'] . '/interface/themes';
	 //echo $themesDir;
	 if (is_dir($themesDir)) {
	 if ($dh = opendir($themesDir)) {
	 while (($file = readdir($dh)) !== false) {
	 if (substr($file, 0, 1) != '.') {
	 $list[] = $file;
	 }
	 }
	 }
	 closedir($dh);
	 }
	 $interface->assign('themeList', $list);

	 $interface->assign('dsn', $this->parseDSN($configArray['Database']['database']));

	 $interface->setTemplate('config-config.tpl');
	 $interface->assign('config', $configArray);
	 }

	 function processConfig()
	 {
	 global $configArray;

	 $configArray['Site']['path']     = $_POST['webpath'];
	 $configArray['Site']['url']      = $_POST['weburl'];
	 $configArray['Site']['local']    = $_POST['localpath'];
	 $configArray['Site']['title']    = $_POST['title'];
	 $configArray['Site']['email']    = $_POST['email'];
	 $configArray['Site']['language'] = $_POST['language'];
	 $configArray['Site']['locale']   = $_POST['locale'];
	 $configArray['Site']['theme']    = $_POST['theme'];

	 $configArray['Index']['engine'] = $_POST['engine'];
	 $configArray['Index']['url']    = $_POST['engineurl'];

	 $configArray['Catalog']['catalog'] = $_POST['ils'];

	 $configArray['Database']['database'] = 'mysql://' . $_POST['dbusername'] . ':' . $_POST['dbpassword'] . '@' . $_POST['dbhost'] . '/' . $_POST['dbname'];

	 $configArray['Mail']['host'] = $_POST['mailhost'];
	 $configArray['Mail']['port'] = $_POST['mailport'];

	 $configArray['BookCovers']['provider']  = $_POST['bookcover_provider'];
	 $configArray['BookCovers']['id']        = $_POST['bookcover_id'];

	 $configArray['BookReviews']['provider'] = $_POST['bookreview_provider'];
	 $configArray['BookReviews']['id']       = $_POST['bookreview_id'];

	 $configArray['LDAP']['host']    = $_POST['ldaphost'];
	 $configArray['LDAP']['port']    = $_POST['ldapport'];
	 $configArray['LDAP']['basedn']  = $_POST['ldapbasedn'];
	 $configArray['LDAP']['uid']     = $_POST['ldapuid'];

	 $configArray['COinS']['identifier'] = $_POST['coinsID'];

	 $configArray['OAI']['identifier']   = $_POST['oaiID'];

	 $configArray['OpenURL']['url']      = $_POST['openurl'];

	 $configArray['EZproxy']['host']     = $_POST['ezproxyhost'];

	 $fileData = '';
	 foreach ($configArray as $name => $section) {
	 $fileData .= "[$name]\n";
	 foreach ($section as $field => $value) {
	 $fileData .= "$field = \"$value\"\n";
	 }
	 }
	 $this->processConfigFile('conf/config.ini', $fileData);
	 }
	 */
	 
	/**
	 * This is lifted, err, adopted from the MDB2::parseDSN function
	 *
	 * Parse a data source name
	 *
	 * Additional keys can be added by appending a URI query string to the
	 * end of the DSN
	 *
	 * The format of the supplied DSN is in its fullest form:
	 * <code>
	 * phptype(dbsyntax)://username:password@protocol+hostspec/database?option=8&another=true
	 * </code>
	 *
	 * Most variations are allowed:
	 * <code>
	 *   phptype://username:password@protocol+hostspec:110//usr/db_file.db?mode=0644
	 *   phptype://username:password@hostspec/database_name
	 *   phptype://username:password@hostspec
	 *   phptype://username@hostspec
	 *   phptype://hostspec/database
	 *   phptype://hostspec
	 *   phptype(dbsyntax)
	 *   phptype
	 * </code>
	 *
	 * @param string Data Source Name to be parsed
	 * @return array an associative array with the following keys:
	 *   + phptype: Database backend used (mysql, odbc, etc.)
	 *   + dbsyntax: Database used with regards to SQL syntax
	 *   + protocol: Communication protocol to use (tcp, unix, etc.)
	 *   + hostspec: Host specification (hostname[:port]
	 *   + database: Database to use on the DBMS server
	 *   + username: User name for login
	 *   + password: Password for login
	 *
	 * @access public
	 * @author Tomas V. V. Cox <cox@idecnet.com>
	 * @author Modified for Vufind by Wayne Graham <wsgrah@wm.edu>
	 */
	function parseDSN($dsn){

		$parsed = array(
            'phptype'  => false,
            'dbsyntax' => false,
            'protocol' => false,
            'hostspec' => false,
            'database' => false,
            'username' => false,
            'password' => false
		);

		// Find phptype and dbsyntax
		if (($pos = strpos($dsn, '://')) !== false) {
			$str = substr($dsn, 0, $pos);
			$dsn = substr($dsn, $pos + 3);
		} else {
			$str = $dsn;
			$dsn = null;
		}

		// Get phptype and dbsyntax
		// $str => phptype(dbsyntax)
		if (preg_match('|^(.+?)\((.*?)\)$|', $str, $arr)) {
			$parsed['phptype']  = $arr[1];
			$parsed['dbsyntax'] = !$arr[2] ? $arr[1] : $arr[2];
		} else {
			$parsed['phptype']  = $str;
			$parsed['dbsyntax'] = $str;
		}

		if (!count($dsn)) {
			return $parsed;
		}

		// Get (if found): username and password
		// $dsn => username:password@protocol+hostspec/database
		if (($at = strrpos($dsn,'@')) !== false) {
			$str = substr($dsn, 0, $at);
			$dsn = substr($dsn, $at + 1);
			if (($pos = strpos($str, ':')) !== false) {
				$parsed['username'] = rawurldecode(substr($str, 0, $pos));
				$parsed['password'] = rawurldecode(substr($str, $pos + 1));
			} else {
				$parsed['username'] = rawurldecode($str);
			}
		}

		// Find protocol and hostspec

		// $dsn => proto(proto_opts)/database
		if (preg_match('|^([^(]+)\((.*?)\)/?(.*?)$|', $dsn, $match)) {
			$proto       = $match[1];
			$proto_opts  = $match[2] ? $match[2] : false;
			$dsn         = $match[3];

			// $dsn => protocol+hostspec/database (old format)
		} else {
			if (strpos($dsn, '+') !== false) {
				list($proto, $dsn) = explode('+', $dsn, 2);
			}
			if (   strpos($dsn, '//') === 0
			&& strpos($dsn, '/', 2) !== false
			&& $parsed['phptype'] == 'oci8'
			) {
				//oracle's "Easy Connect" syntax:
				//"username/password@[//]host[:port][/service_name]"
				//e.g. "scott/tiger@//mymachine:1521/oracle"
				$proto_opts = $dsn;
				$dsn = substr($proto_opts, strrpos($proto_opts, '/') + 1);
			} elseif (strpos($dsn, '/') !== false) {
				list($proto_opts, $dsn) = explode('/', $dsn, 2);
			} else {
				$proto_opts = $dsn;
				$dsn = null;
			}
		}

		// process the different protocol options
		$parsed['protocol'] = (!empty($proto)) ? $proto : 'tcp';
		$proto_opts = rawurldecode($proto_opts);
		if (strpos($proto_opts, ':') !== false) {
			list($proto_opts, $parsed['port']) = explode(':', $proto_opts);
		}
		if ($parsed['protocol'] == 'tcp') {
			$parsed['hostspec'] = $proto_opts;
		} elseif ($parsed['protocol'] == 'unix') {
			$parsed['socket'] = $proto_opts;
		}

		// Get dabase if any
		// $dsn => database
		if ($dsn) {
			// /database
			if (($pos = strpos($dsn, '?')) === false) {
				$parsed['database'] = $dsn;
				// /database?param1=value1&param2=value2
			} else {
				$parsed['database'] = substr($dsn, 0, $pos);
				$dsn = substr($dsn, $pos + 1);
				if (strpos($dsn, '&') !== false) {
					$opts = explode('&', $dsn);
				} else { // database?param1=value1
					$opts = array($dsn);
				}
				foreach ($opts as $opt) {
					list($key, $value) = explode('=', $opt);
					if (!isset($parsed[$key])) {
						// don't allow params overwrite
						$parsed[$key] = rawurldecode($value);
					}
				}
			}
		}

		return $parsed;
	}
	 
	function getAllowableRoles(){
		return array('opacAdmin');
	}
}